examples/shopping_ads/add_shopping_smart_ad.rb

```diff
@@ -0,0 +1,274 @@
+    # of product groups in the API and allow you to bid on the chosen group or
+    # exclude a group from bidding.
+    # This method creates an ad group criterion containing a listing group.
+    add_default_shopping_listing_group(client, customer_id, ad_group_name)
+  end
+end
+
+def add_campaign_budget(client, customer_id)
+
+  operation = client.operation.create_resource.campaign_budget do |budget|
+    budget.name = "Interplanetary Budget ##{(Time.new.to_f * 1000).to_i}"
+    budget.delivery_method = :STANDARD
+    budget.amount_micros = 500_000
+    # Budgets for Smart Shopping campaigns cannot be shared.
+    budget.explicitly_shared = false
+  end
+
+  service = client.service.campaign_budget
+  response = service.mutate_campaign_budgets(
+    customer_id: customer_id,
+    operations: [operation],
+  )
+
+  budget_name = response.results.first.resource_name
+
+  puts "Added a budget with resource name: #{budget_name}"
+
+  budget_name
+end
+
+def add_smart_shopping_campaign(
+    client, customer_id, budget_name, merchant_center_id)
+
+  operation = client.operation.create_resource.campaign do |campaign|
+    campaign.name = "Interplanetary Cruise Campaign ##{(Time.new.to_f * 1000).to_i}"
+
+    # Configures settings related to shopping campaigns including
+    # advertising channel type, advertising sub-type and shopping setting.
+    campaign.advertising_channel_type = :SHOPPING
+    campaign.advertising_channel_sub_type = :SHOPPING_SMART_ADS
+    campaign.shopping_setting = client.resource.shopping_setting do |shopping_setting|
+      shopping_setting.merchant_id = merchant_center_id
+      # Sets the sales country of products to include in the campaign.
+      # Only products from Merchant Center targeting this country will appear in the
+      # campaign.
+      shopping_setting.sales_country = "US"
+    end
+
+    # Recommendation: Set the campaign to PAUSED when creating it to prevent
+    # the ads from immediately serving. Set to ENABLED once you've added
+    # targeting and the ads are ready to serve.
+    campaign.status = :PAUSED
+
+    # Bidding strategy must be set directly on the campaign.
+    # Setting a portfolio bidding strategy by resource name is not supported.
+    # Maximize conversion value is the only strategy supported for Smart Shopping
+    # campaigns.
+    # An optional ROAS (Return on Advertising Spend) can be set for
+    # MaximizeConversionValue.
+    # The ROAS value must be specified as a ratio in the API. It is calculated by dividing
+    # "total value" by "total spend".
+    # For more information on maximize conversion value, see the support article:
+    # http://support.google.com/google-ads/answer/7684216.
+    campaign.maximize_conversion_value = client.resource.maximize_conversion_value do |m|
+      m.target_roas = 3.5
+    end
+
+    campaign.campaign_budget = budget_name
+  end
+
+  service = client.service.campaign
+  response = service.mutate_campaigns(
```

Comment:
Just make the service inline if you only use it once: response = client.service.campaign.mutate_campaigns(...)

Reply:
done.